### ISSUE TITLE
Towards round trip data flow analyis using the web editor

### ISSUE THREAD

**Issue by sebinside** [DataFlowAnalysis/DataFlowAnalysis#175]
Demonstration will be easier than ever with the semantic preserving conversion of DFD to PCM of #170. However, @Nicolas-Boltz and I came up with an even better version that uses our current analysis and display capabilities more efficiently.

## 1. Display analysis results in web editor

In the first step, we should use the node annotations of the web editor (see https://github.com/DataFlowAnalysis/WebEditor/pull/42) in the converter from DFD to Web JSON. This does not only include the violations but also the propagated labels (which is, in a way, the requirement that Nicolas had in the last meeting, but this time without any hacks). I attached a simple JSON file. The (made up) violation is shown in red, hovering over the user node also shows the propagated labels.

The good thing is, that these annotations are really simple, e.g., in the attached example:

```
"annotation": {
                    "message": "Propagated labels: data.personal",
                    "icon": "tag"
                }
```

and 

```
"annotation": {
                    "message": "Violation: Constraint 42 has been violated.",
                    "icon": "bolt",
                    "color": "#ff0000"
                }
```

We could even argue that this should be the default behavior of the converter (which might be possible to disable via a flag) because this is the most user friendly way how to display analysis results.

[analysis-result.json](https://github.com/user-attachments/files/16015578/analysis-result.json)

## 2. Round trip between editor and analysis

There is another nice feature in the current version of the web editor (which is also documented [here](https://github.com/DataFlowAnalysis/WebEditor/pull/42); really, really good work @hlxid!):
```
"editorMode": "annotated"
```

This feature makes the web editor read-only and the user has to click a button that enables changing the diagram again but removes all annotations. This made @Nicolas-Boltz and myself thinking and we came up with an idea which would give us the most brilliant demonstration capabilities ever (without jokes, for real, this was like a big vision two years ago, and now it becomes reachable).

Executed locally, we (probably) could force the web editor to write to and load from a file (or get the json of the model from any other source like a locally hosted service). By having an eclipse instance (headless, or a test case) that watches this file for changes (or provides the endpoint for the json data), we could run the analysis if the diagram is changed, display the changes a few seconds after this, allow the editor to make changes, and start over again. This way, we don't have to update the core analysis, but users will only see the web editor front end. And *drumroll* because of #170, this would even work with PCM models after conversion. Throw this into a docker container, and you have reached demonstrator end game. In the future, we could then even enable writing DSL constraints web based (but that's really future work).

## TL;DR

With the upcoming 3.0.0 version of the data flow analysis, we already provide a comprehensive, well-published analysis framework. However, changes in the proposed direction would not only simplify demonstrations (by a lot) but also reach the vision of light-weight and future-proof architectural analysis. We are close!

**Comment by sebinside** [comment]
(This issue is also a follow-up to #129)

**Comment by uuqjz** [comment]
Currently, the editor has some semantics that cannot be represented in a DFD. Hence complicating the round-trip idea
E.g.:
`set Encryption.Encrypted = request.Encryption.Encrypted && data.Encryption.Encrypted`
`set Encryption.Encrypted = request.Encryption.Encrypted && data.Location.Somewhere`
These two terms are legal in the web editor but cannot be correctly represented in a DFD

**Comment by sebinside** [comment]
Discussion:
* Result display should be part of the converter
* A solution to @uuqjz comment could be a reference to the (named) input pin in the assignment
* In that direction, we could also include selected forwarding, i.e., only forwarding all labels of a type
* Nils has to do a lot. See #192 and #193
* A solution could also be to skip the DFD entirely and write a TFGFinder for WebDFD

**Comment by sebinside** [comment]
New assignment format:

```
Old
Forwarding({data})
Assignment({};TRUE;{Encryption.Encrypted})

New
forward data,y,z
set Encryption.Encrypted,A.B
unset Encryption.Encrypted
assign Encryption.Encrypted,A.B if TRUE
assign Encryption.Encrypted,A.B if TRUE from data
```

**Comment by sebinside** [comment]
Constraint Format Inspiration: 
[https://sebastianhahner.de/publications/2024/BoltzHahner2024_AnExtensibleFrameworkForArchitectureBasedDataFlowAnalysisForInformationSecurity.pdf](https://sebastianhahner.de/publications/2024/BoltzHahner2024_AnExtensibleFrameworkForArchitectureBasedDataFlowAnalysisForInformationSecurity.pdf), Listing 1
[https://sebastianhahner.de/publications/2021/Hahner2021_ModelingDataFlowConstraintsForDesignTimeConfidentialityAnalyses.pdf](https://sebastianhahner.de/publications/2021/Hahner2021_ModelingDataFlowConstraintsForDesignTimeConfidentialityAnalyses.pdf), Listing 1 + 2

**Comment by 01Parzival10** [comment]
> New assignment format:
> 
> ```
> Old
> Forwarding({data})
> Assignment({};TRUE;{Encryption.Encrypted})
> 
> New
> forward data,y,z
> set Encryption.Encrypted,A.B
> unset Encryption.Encrypted
> assign Encryption.Encrypted,A.B if TRUE
> assign Encryption.Encrypted,A.B if TRUE from data
> ```

Should Set/UnsetAssignment inherit from AbstractAssignment? If so then what of the inherited input Pins

**Comment by 01Parzival10** [comment]
I performed the naming changes and added the new Assignments in:

- DataFlowAnalysis
- Converter
- ExampleModels
- DFD-Metamodel

I pushed it all to a 4.0 branch

**Comment by 01Parzival10** [comment]
This batch program "fixes" all diagrams in the folder and subfolder

```batch 
@echo off
setlocal enabledelayedexpansion

REM Set the directory to process (current directory by default)
set "targetDir=%~dp0"

REM Iterate through all files in the directory and subdirectories
for /r "%targetDir%" %%f in (*.dataflowdiagram *.datadictionary) do (
    REM Create a temporary file to store updated content
    set "tempFile=%%~dpnxf.temp"

    REM Read the file line by line and replace "behaviour" with "behavior"
    for /f "delims=" %%l in (%%f) do (
        set "line=%%l"
        set "line=!line:behaviour=behavior!"
        echo !line! >> "!tempFile!"
    )

    REM Replace the original file with the updated file
    move /y "!tempFile!" "%%f" >nul
)

echo Replacement completed.
pause

**Comment by 01Parzival10** [comment]
> > New assignment format:
> > ```
> > Old
> > Forwarding({data})
> > Assignment({};TRUE;{Encryption.Encrypted})
> > 
> > New
> > forward data,y,z
> > set Encryption.Encrypted,A.B
> > unset Encryption.Encrypted
> > assign Encryption.Encrypted,A.B if TRUE
> > assign Encryption.Encrypted,A.B if TRUE from data
> > ```
> 
> Should Set/UnsetAssignment inherit from AbstractAssignment? If so then what of the inherited input Pins

Move input Pins from AbstractAssignment to Forwarding Assignments and Assignments
